### PR TITLE
lxd-ui: 0.2 -> 0.4

### PR DESCRIPTION
**Pull request by jnsgruk**
## Description of changes

Updates the LXD UI to the latest release (0.4) and also fixes the package. This has been broken for a little while - I've done some experimentation and I can't get the produced bundle to work at all when `mkYarnPackage` is used. This approach seems to produce a working JS bundle at the end.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by adamcstephens** [comment]
> also fixes the package

How was it broken? There is a test for the UI that is not failing. Can it be improved?

**Comment by jnsgruk** [comment]
The answer to that is: I don't really know. The compiled bundle throws an error that's indecipherable:

image

I've spent a few hours tinkering with the various phases in `mkYarnPackage` to try and make it work properly, but to no avail. I can't reproduce the failure if I build outside of the nix build process with the exact same sequence of commands that `mkYarnPackage` and `fetchYarnDeps` uses, so I'm at a bit of a loss (as are the upstream devs!)

**Comment by adamcstephens** [comment]
Yeah, I'm not doubting it's broken nor suggesting the PR changes are bad. I was just wondering if we can test this better. Unfortunately, the current nixos test does not actually validate the bundled javascript. The UI test only validates that LXD is correctly serving the UI HTML file.

**Comment by jnsgruk** [comment]
Ah I see! :) Yes, it had crossed my mind that the current test is not great, I've added a selenium test for a bit more rigour here :)

File: nixos/tests/lxd/ui.nix [hunk 1]
```diff
@@ -11,9 +11,37 @@ import ../make-test-python.nix ({ pkgs, lib, ... }: {
...
+          options.add_argument("--headless")
+          service = webdriver.FirefoxService(executable_path="${lib.getExe pkgs.geckodriver}")  # noqa: E501
+
+          driver = webdriver.Firefox(options=options, service=service)
+          driver.implicitly_wait(10)
+          driver.get("https://localhost:8443/ui")
+
+          wait = WebDriverWait(driver, 60)
+
+          assert len(driver.find_elements(By.CLASS_NAME, "l-application")) > 0
```
Comment (jnsgruk) [inline nixos/tests/lxd/ui.nix:35]: These elements only appear on the page if the JavaScript bundle inits itself properly and renders the app.

File: pkgs/tools/admin/lxd/ui.nix [hunk 2]
```diff
@@ -1,35 +1,61 @@
...
   pname = "lxd-ui";
-  version = "0.2";
+  version = "0.4";
 
   src = fetchFromGitHub {
     owner = "canonical";
     repo = "lxd-ui";
     rev = "refs/tags/${version}";
-    sha256 = "sha256-DygWNktangFlAqinBm6wWsRLGmX6yjhmRJ2iU0yjcgk=";
+    sha256 = "sha256-l9Fp/Vm7NxMCp5QcM8+frFyfahhPG7TyF6NhfU1SEaA=";
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/ui.nix:18]:
```suggestion
    hash = "sha256-l9Fp/Vm7NxMCp5QcM8+frFyfahhPG7TyF6NhfU1SEaA=";
```

File: pkgs/tools/admin/lxd/ui.nix [hunk 3]
```diff
@@ -1,35 +1,61 @@
...
   };
 
-  packageJSON = ./package.json;
   offlineCache = fetchYarnDeps {
     yarnLock = "${src}/yarn.lock";
-    sha256 = "sha256-B1SVCViX1LEFoBLMdFk9qaoayku7Y+zU5c4JEJkLmwE=";
+    sha256 = "sha256-R6OeaBC6xBHa229YGyc2LDjjenwvS805PW8ueU/o99I=";
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/lxd/ui.nix:23]:
```suggestion
    hash = "sha256-R6OeaBC6xBHa229YGyc2LDjjenwvS805PW8ueU/o99I=";
```

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/tools/admin/lxd/ui.nix (hunk 3 above)
Comment (jnsgruk) [reply pkgs/tools/admin/lxd/ui.nix]: Done!

File: pkgs/tools/admin/lxd/ui.nix (hunk 2 above)
Comment (jnsgruk) [reply pkgs/tools/admin/lxd/ui.nix]: Done!

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**